backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,358 @@
+      // Returning the existing assignment if some backend is something already assigned to this process group and it is not defunct
+      if(logger.isDebugEnabled()) {
+        logger.debug(String.format("process_group=%s, existing backend=%s, available",
+            ProtoUtil.processGroupCompactRepr(processGroup), backendAssociation));
+      }
+      result.complete(backendAssociation);
+    } else {
+      vertx.executeBlocking(future -> {
+        /**
+         * Since a backend assignment and possible de-assignment is required, acquiring a lock to avoid race conditions. Examples:
+         * => two recorders with same process group getting different backends assigned
+         * => available backend queue seen as empty by some of the requests because assignments of all available backends were being updated in other requests.
+         *    this can arise because during assignment, backend is dequeued from available backend queue and enqueued again post assignment
+         * => competing zookeeper requests to de-assign a defunct backend
+         */
+        if(logger.isDebugEnabled()) {
+          logger.debug(String.format("process_group=%s, existing backend=%s, defunct or null",
+              ProtoUtil.processGroupCompactRepr(processGroup), backendAssociation));
+        }
+        try {
+          boolean acquired = backendAssignmentLock.tryLock(2, TimeUnit.SECONDS);
+          if (acquired) {
+            try {
+              //Lookup existing backend association again after acquiring lock to avoid race conditions
+              String existingBackendAssociation = processGroupToBackendLookup.get(processGroup);
+              if (existingBackendAssociation == null) {
+                //This is a new process group and no backend has been assigned to this yet
+                BackendDetail availableBackend = getAvailableBackendFromPriorityQueue();
+                try {
+                  if (availableBackend == null) {
+                    future.fail("No available backends are known to leader, cannot assign one to process_group=" + processGroup);
+                  } else {
+                    try {
+                      associateBackendWithProcessGroup(availableBackend, processGroup);
+                      if (logger.isDebugEnabled()) {
+                        logger.debug(String.format("process_group=%s, new backend=%s",
+                            ProtoUtil.processGroupCompactRepr(processGroup), availableBackend.getBackendIPAddress()));
+                      }
+                      future.complete(availableBackend.getBackendIPAddress());
+                    } catch (Exception ex) {
+                      future.fail(String.format("Cannot persist association of backend=%s with process_group=%s in zookeeper",
+                          availableBackend.getBackendIPAddress(), processGroup));
+                    }
+                  }
+                } finally {
+                  if (availableBackend != null) {
```

Comment:
Another simplification, because this is conditional, it'll be more readable if we fold this inside else (we have a try-catch anyway).